### PR TITLE
!!! BUGFIX: Prevent not fully translated content from getting lost in move operations

### PR DESCRIPTION
**Pull request by Sebobo**
This resolves the issue that moving a collection with subnodes in dimension value A doesn’t update the parent path of nodes in dimension value B if the collection itself does not exist in language B. The result would be that the content in B is now disconnected from the tree and not reachable anymore.

This slightly changes the behavior in existing projects, but the change keeps content from getting lost and is therefore essential for every Neos project with dimensions and fallbacks.
The new behaviour is also more in line with the default behaviour of the new Content Repository in Neos 9.

Resolves: #3340

**Upgrade instructions**

The `aggregate` mode of the collection will now enforce that the content is moved in all dimensions at the same time to the same parent path. This is different than from the behaviour before. 

If the new behaviour is a problem in your project, you can revert the setting with the following yaml nodetype configuration:

```yaml
Neos.Neos:ContentCollection':
  aggregate: false
```

### CODE REVIEW

**Comment by Sebobo** [comment]
😔 There seems to be an issue that a node cannot be moved even if it has no translations if it itself is an aggregate and is moved into another one. In the Neos.Demo this would be moving a two column content inside a column of another two column content. I'm investigating.

**Comment by Sebobo** [comment]
@kitsunet we have a problem f.e. with column elements when the children are content collections but the element itself isn't. Because in this case the children are moved in `setPathInternal` before their parent, but with `setPathInternalForAggregate` first the parent and then the children...
So the first case throws an error.

**Comment by Sebobo** [comment]
I now adjust the behaviour so that in both cases the parent is moved **before** the children, which fixes the issue.

**Comment by Sebobo** [comment]
Now tests are failing as expected. Will check next week.

**Comment by bwaidelich** [comment]
Just checking, is the `aggregate` flag really suitable for content nodes?

**Comment by Sebobo** [comment]
@bwaidelich afaik the idea of the aggregate is, to move independent (not auto-created) child nodes with a parent node to prevent them from getting lost and doomed. That is also what I understood from Bernhard the behaviour of the new CR.

I fixed the tests.

File: Neos.ContentRepository/Classes/Domain/Model/Node.php [hunk 1]
```diff
@@ -237,14 +237,15 @@ protected function isNodePathAvailable(string $path): bool
     protected function setPathInternal(string $destinationPath, bool $recursiveCall): array
     {
         $originalPath = $this->nodeData->getPath();
+        $childNodes = $this->getChildNodes();
+
+        $this->moveNodeToDestinationPath($this, $destinationPath);
```
Comment (kitsunet) [inline Neos.ContentRepository/Classes/Domain/Model/Node.php:242]: Wait but this moves the parent before the children, why do we need this (I fear this might have unintended side effects further down the chain as all the children will no longer have a connection to root)?

Comment (Sebobo) [reply Neos.ContentRepository/Classes/Domain/Model/Node.php:242]: When an aggregate is moved there is a check that the children moved to an existing parent, so the parent and children are moved together.

This is the other way around for non-aggregates as the children are moved first (for unknown reasons to me) and then the parent which will make the integrity check fail when an aggregate is inside a non aggregate as the children are then disconnected until the parent is moved. IMO this could maybe explain some other issues we had in the past of disconnected children. See the problem described in https://github.com/neos/neos-development-collection/issues/3340#issuecomment-1673068873

Comment (mhsdesign) [reply Neos.ContentRepository/Classes/Domain/Model/Node.php:242]: Hmm sorry i cant follow your argument @Sebobo ... could you please elaborate further?

Comment (Sebobo) [reply Neos.ContentRepository/Classes/Domain/Model/Node.php:242]: Discussed via Slack and agreed upon that we need to add some tests

**Comment by mhsdesign** [comment]
do we still want this to go in 8.3 as bugfix?


and is this possibly related? https://github.com/neos/neos-development-collection/pull/3241

**Comment by Sebobo** [comment]
Would be great to have this, as every project with dimensions could loose content. 8.3 would now also be fine with me.

**Comment by mhsdesign** [comment]
It would be interesting to know what the neos 9 behaviour is and having a behat feature for that. 
And this is change is definitely breaking as it doesnt allow to scatterly move a content collection

EDIT: after talking to bernhard we are sure that this bug should not occur in Neos 9. I would suggest NOT to merge this as a bugfix (as its super breaking) ... everyone who needs this can use composer patches (`https://github.com/neos/neos-development-collection/pull/4436.patch`)

**Review by mhsdesign — CHANGES_REQUESTED** [review]
As by my last comment i think we should not merge this or at least synchronise with the 9.0 people if this fix makes sense in the long term.

cc @bwaidelich @nezaniel @skurfuerst

**Comment by Sebobo** [comment]
> It would be interesting to know what the neos 9 behaviour is and having a behat feature for that. And this is change is definitely breaking as it doesnt allow to scatterly move a content collection

What does that mean?
Do you have a test case that breaks?

**Comment by nezaniel** [comment]
First: Thanks for the detailed description, this was a scary, yet fun detective endeavour.

Caution: may contain walls of text and traces of graph theory.

TLDR: We have to decide on a strategy to use when moving content collection nodes and either go for
* aggregate:true, which is a breaking bugfix because it changes the fundamental behavior among variants
* fixing move node in Neos 8 without changing the behavior, which is possible by backporting the scatter or gatherVirtualSpecializations move strategy behavior from Neos 9

**What's actually going on here**
I'll use the Neos 9 terminology and graph model for this topic, because we now have a language to describe the behavior even in Neos <=8.

The situation before the fateful move operation is as follows:

LostUntranslatedContent-BeforeMove

For those unfamiliar with the content graph model:
* everything red is German, either as a node (which language is the content in) or as an edge (what is visible if I look at the site in this language)
* everything violet is English, either as a node with edge (an actual translation) or as a fallback edge connecting a German node, describing the fallback mechanism
* the node labels describe the node identifiers, the edge labels describe the node names, this means the node's path can be read from the edges (this is the way it really works in 9, in 8 the complete path is actually written to the node, which is the root of our problem here as we will soon see)

Properties are written in [] but are not super important right now. The only important thing here is that (implicit) translations of text1 and text2 have been created by changing the text property.

Up until now, it doesn't really matter whether we have edges or paths in nodes, Neos 8 and 9 handle this pretty similar.
But now we move a node, which is where the graph model shines and we have to decide how to deal with node paths in Neos 8.

In Neos 9, we describe four different strategies on how to do this (with one still being in development):
* scatter: This means that only the selected variant is moved, all other, including virtual specializations (the fallback), is left behind
* gatherVirtualSpecializations: This means that the selected variant and all virtual specializations are moved. In this case this also affects English because the ColGrid has not been translated. Had it been translated it would have been ignored. 
* gatherSpecializations: This means that the selected variant and all specializations are moved. In this case this would affect English whether the ColGrid has been translated or not
* gatherAll: This means that all variants (specializations, peers and generalizations) are affected

In Neos 8, aggregate:true resembles the gatherAll strategy, while aggregate:false is in a hard place between scatter and gatherVirtualSpecializations and thus broken by design. It has to choose to be one or the other, as we'll see soon.

Now, let's move the Node in Neos 9...
... with the scatter strategy
LostUntranslatedContent-Neos9Scatter
This is pretty simple, the German edge is reconnected and everything English stays the same.

We can also choose one of the other strategies, which will also affect English:
LostUntranslatedContent-Neos9Gather
Same story, reconnect two edges and be happy.

This doesn't directly help for Neos 8, but it helps showing what choices we have (and have to take), what the desired outcome is and that this issue just evaporates when upgrading to Neos 9.

But now let's try the same with aggregate:false in Neos 8:
LostUntranslatedContent-Neos8

What we see here is a graph representation of orphan nodes. The bare edges represent the path stored in the NodeData records with the parent segments pointing nowhere. So what happened?

When moving a Node with aggregate:false, Neos8- attempts something similar to the scatter strategy; via recursive Node::getChildNodes, all descendants' paths are adjusted. That means that the content collection nodes are affected as well, since they can be resolved via getChildNodes. But when they are moved, so is their English virtual variant (accidentally?), since the fallback calculation is done at read time. When we reach the Text nodes, only the non-translated ones are moved because the translated ones are not resolved via getChildNodes.
In Neos 9 terms, we apply a mixture of scatter (for translated nodes) and gatherVirtualSpecializations (for non-translated nodes) which is doomed to fail.

Side note: The only reason the German nodes don't appear as English at the new location is that there are (orphaned) English nodes with the same identifier. 

So what can we do instead?

Option 1: Switch to aggregate:true
As said before, this is breaking for non-documents as they now behave differently. It would have to be applied to everything that can have descendants, because this is a general descendant issue when fallbacks are involved.
The result would be as follows:
LostUntranslatedContent-Neos8-gatherAll

Option 2: Switch to the scatter strategy
To do this, we need to really leave the English content collections behind while moving. But since paths are stored in the nodes, this can only be done by translating the to-be-moved node (and their non-translated descendants) to English first:
LostUntranslatedContent-Neos8-scatter

Option 3: Switch to the gatherVirtualSpecializations strategy
To do this, we have to move the translated English descendants as well, because language:English is affected by the operation in general.
(note to whoever might implement this: be wary of translated content collections. If their translations are not to be moved (since they are not virtual) and they have non-translated descendants, those have to be translated before move, see Option 2)
In our diagram, the result is the same as Option 1, just that other, unrelated language variants are unaffected.

Either way, parents must be moved before their children are, because of the constraint checks imposed since Neos 8 (and common sense, really). 

The challenge for Options 2 and 3 is the determination of the affected dimension space points (in this case: language values). We need to know the coverage of the to-be-moved node, i.e. in which languages etc. it is visible due to the fallback mechanism.
There is a standalone version of the DimensionSpace available (https://github.com/neos/content-repository-dimensionspace) that could come in handy.
Otherwise a look at how Neos visualizes the dimensions (https://github.com/neos/neos-development-collection/blob/8.3.12/Neos.Neos/Classes/Presentation/Model/Svg/InterDimensionalFallbackGraph.php) might help.
Once we know which variants to handle, we can apply the necessary operations (translation in the scatter case or path change and translation in the gatherVirtualSpecializations case) on them by switching the ContentContext and then traversing again.

I hope this clears up stuff (and not too much), any questions? :)
